Ray step4

torch-ray opened the pull request:
## Step4 작업목록
- [x] 앱 생명주기관리 학습
- [x] 앱 종료 후 객체속성 저장
- [x] 앱 재실행시 마지막 재고상태 복원
- [x] AppDelegate 전역변수 학습
- [x] UserDefault 라이브러리 사용
- [x] 실행화면 캡쳐

## 학습키워드
- AppDelegate
- SceneDelegate

## 고민과 해결
- AppDelegate에 객체 인스턴스를 생성 후, ViewController에서 접근 시도
- 원하는대로 코드가 작동하지 않아 확인해보니 앱딜리의 foreground 및 background의 print함수가 작동하지 않는 것을 발견
- iOS 13부터 AppDelegate와 SceneDelegate의 역할이 분리
- SceneDelegate에 전역변수를 생성하여 접근하는 식으로 변경
- foreground 및 background의 print함수는 정상 작동
- ViewController에서 접근시 무조건 nil값 반환</br>
[고민과 해결 정리](https://velog.io/@torch-ray/iOSlearn1)

## 부연설명
- (생략)

## 질문거리
- AppDelegate와 SceneDelegate를 핵심적으로 학습하였습니다. 하지만, 아직 ViewController에서 AppDelegate으로의 접근과 SceneDeleagate로의 접근이 어떻게 다르길래 SceneDelegate에서의 전역변수로 접근시에는 무조건 nil값을 반환하는지 모르겠습니다! 그저 AppDelegate가 앱 전체적인 관리, SceneDelegate는 UI LifeCycle에서의 관리의 차이 때문이라고 짐작만하고 있습니다. 해당 내용관련하여 추가적으로 학습을 진행할 수 있는 힌트가 있을까요?

torch-ray committed:
Codable+UserDefaults활용하여 저장시도하였으나, 실패.. 에러는 없는데 앱종료하고 켜면 데이터 저장 안됨

torch-ray committed:
UserDefaults활용하여 앱종료후에도 데이터 저장되도록 코딩 구현

토요일, 일요일 내내 SceneDelegate에서 삽질하다가 nil값만 2000번 반환받고 멘탈붕괴됐다가 AppDelegate에서 전역변수(?)로 VendingMachine선언하고 해결했습니다.

기본적으로 앱 종료될 때 앱딜리에선는 DidEnterBackground 코드는 작동하지 않고 신딜리에서만 작동하길래 신딜리에서 작업했는데 이게 화근이 됐습니다.

앱딜리게이트 신딜리게이트를 공부하게된 좋은 계기였습니다. 둘 사이의 차이점은 정리해서 블로그에 기록할 예정입니다.

지나치게 오랜기간 삽질하다보니 git으로 기록하는 것도 잊었습니다.. 추후 작업시에는 삽질을 하더라도 git으로 중간중간 과정을 기록할 예정입니다.

torch-ray committed:
IBOutlet Collection으로 UILabel통합

torch-ray committed:
ReadMe.md 파일 수정

godrm commented on VendingMachineApp/VendingMachineApp/Model/Beverage.swift:
```diff
@@ -1,6 +1,6 @@
 import Foundation
 
-class Beverage {
+class Beverage: Codable {
```
https://medium.com/@codesquad_yoda/codable-vs-nscoding-차이점-4b47e240c0b8
이 글과 영상을 한 번 보세요. 도움이 될 겁니다.

torch-ray replied:
해당 내용 학습 후 NSCoding채택으로 수정하였습니다.

godrm commented on VendingMachineApp/VendingMachineApp/Model/Beverage.swift:
```diff
@@ -22,11 +33,35 @@ class Beverage {
+    }
+    
     func availableForBeverage() -> Bool {
         let machine = VendingMachine()
-        return machine.insertedMoney.compareMoney(to: price) == true ? true : false
+        return machine.insertedMoney.compareMoney(to: price) == true
     }
```
이 메소드는 나쁜 사례입니다.
왜냐하면 하위 모듈인 음료 객체가 상위 모듈인 VendingMachine을 직접 참조하거나 인스턴스를 생성하기 때문입니다.
음료수가 자판기 구체 타입에 의존하도록 만들면 안됩니다.

torch-ray replied:
참고하여 코드 수정하였습니다 감사합니다!

godrm commented on VendingMachineApp/VendingMachineApp/Model/Beverage.swift:
```diff
@@ -22,11 +33,35 @@ class Beverage {
     }
-    
+
     func afterBuyingBeverage() {
         let machine = VendingMachine()
         machine.insertedMoney.afterBuyingProduct(minus: price)
```
이 함수도 위와 같은 이유료 나쁜 구조입니다. 이런 타입을 쓰는 경우는 음료수가 자판기보다 상위 모듈인 경우만 쓰는 겁니다.

godrm commented on VendingMachineApp/VendingMachineApp/Model/VendingMachineData.swift:
```diff
@@ -0,0 +1,22 @@
+    
+    private let key = "encodedData"
+    
+    func loadedData() -> VendingMachine? {
+        guard let data = UserDefaults.standard.data(forKey: key) else {
+            return nil
+        }
+        
+        guard let vendingMachine = try? PropertyListDecoder().decode(VendingMachine.self, from: data) else {
+            return nil
+        }
+        return vendingMachine
+    }
+    
+    func savedData(_ vendingMachine: VendingMachine) {
+        let encodedData = try? PropertyListEncoder().encode(vendingMachine)
+        UserDefaults.standard.setValue(encodedData, forKey: key)
+    }
```
자판기 객체를 만들어주는 상위 타입이 생겼네요. 
이렇게 생성하는 타입이 있으면 VendingMachine을 직접 init()하는 경우와 헷갈리거나 실수할 여지가 많아집니다. 
AppDelegate나 ViewController에서 VendingMachine 인스턴스를 만들 때 여기서 만들어진 것만 사용하도록 해야합니다. 엄청 귀찮은 일이죠. 이런 경우는 인스턴스가 하나만 만들어지도록 관리하는 게 좋습니다. 
VendingMachine가 여러개 생기지 않도록 주의하세요. 그럴려면 VendingMachine.init() 자체에 접근하지 못하도록 하는 방법을 고려해보세요.

torch-ray replied:
싱글톤으로 코드 구현하였습니다!

godrm commented on VendingMachineApp/VendingMachineApp/Model/VendingMachineData.swift:
```diff
@@ -0,0 +1,22 @@
+        return vendingMachine
+    }
+    
+    func savedData(_ vendingMachine: VendingMachine) {
```
savedData()는 과거형인 이유가 있나요?

torch-ray replied:
저장된 데이터/ 불러온 데이터 로 표기하였는데, 생각해보니 과거형일 이유가 없어서 수정하였습니다!

godrm commented on VendingMachineApp/VendingMachineApp/Model/VendingMachineData.swift:
```diff
@@ -0,0 +1,22 @@
+    
+    private let key = "encodedData"
+    
+    func loadedData() -> VendingMachine? {
+        guard let data = UserDefaults.standard.data(forKey: key) else {
+            return nil
+        }
+        
+        guard let vendingMachine = try? PropertyListDecoder().decode(VendingMachine.self, from: data) else {
+            return nil
+        }
+        return vendingMachine
+    }
```
dataManager가 인스턴스로 machine을 갖고 있지도 않는데 loadedData() 나 savedData() 는 응집도가 높은 타입일까요? 여기 있지 않아도 된다는 의미일까 고민해보세요.

godrm commented on VendingMachineApp/VendingMachineApp/SceneDelegate.swift:
```diff
@@ -18,17 +18,16 @@ class SceneDelegate: UIResponder, UIWindowSceneDelegate {
 
     func sceneDidEnterBackground(_ scene: UIScene) {
-
+        myDelegate?.dataManager.savedData(myDelegate!.machine)
```
myDelegate에게 myDelegate가 가진 것을 넘겨주는 이 메소드 호출 구조가 어색합니다.

godrm commented on VendingMachineApp/VendingMachineApp/VendingMachineViewController.swift:
```diff
@@ -2,26 +2,23 @@ import UIKit
     override func viewDidLoad() {
         super.viewDidLoad()
+        let myDelegate = UIApplication.shared.delegate as! AppDelegate
+        vendingMachine = myDelegate.machine
```
이렇게 하면 AppDelegate에 있는 machine이 소스 레퍼런스가 되고 여기는 참조가 됩니다.
만약 AppDelegate에 있는 machine 값이 바뀌면 여기는 어떻게 될까요?

torch-ray replied:
앱딜리의 machine값이 바뀌면 vendingMachine값도 같이 변한다고... 생각했습니다...?

godrm commented on VendingMachineApp/VendingMachineApp/VendingMachineViewController.swift:
```diff
@@ -33,66 +30,73 @@ class VendingMachineViewController: UIViewController {
-        stockInfoOfSprite.text = "\(sprite)개"
-        stockInfoOfTOP.text = "\(top)개"
-        stockInfoOfCantata.text = "\(cantata)개"
+        let chocoMilk = vendingMachine.stockOf(beverage: factory.chocoMilk)
```
여기 넘어가는 `factory.chocoMilk`가 인스턴스인가요?
약간 비유적으로 표현하자면 음료수별 타입을 가져오는게 아니라, 공장에서 초코밀크를 가져와서 자판기에 보여주면 무엇을 넘겨준건가요? 다시 chocoMilk 인스턴스를 넘겨준건가요?
이 문장만 보면 `개수를 넘겨주는 거구나` 동작을 유추하기 어렵습니다. 왜 그럴까요?

torch-ray replied:
편의 이니셜라이저 추가하여 코드 수정하였습니다!

godrm commented on VendingMachineApp/VendingMachineApp/VendingMachineViewController.swift:
```diff
@@ -33,66 +30,73 @@ class VendingMachineViewController: UIViewController {
+        let coke = vendingMachine.stockOf(beverage: factory.coke)
+        let sprite = vendingMachine.stockOf(beverage: factory.sprite)
+        let top = vendingMachine.stockOf(beverage: factory.top)
+        let cantata = vendingMachine.stockOf(beverage: factory.cantata)
```
그리고 요구사항은 음료 종류별로 카운트를 해서 한꺼번에 넘겨주는 게 아니었던가요?

torch-ray replied:
stockOf메서드는 현재 자판기의 재고중에서 파라미터로 넘겨주는 음료의 수를  filter로 걸러서 해당 음료의 수만 넘겨주는 방식입니다! 음료를 종류별로 카운트해서 한꺼번에 넘겨주는 방식이라면 현재 자판기의 모든 재고를 리턴하는 것 말씀하시는건가요? 어떻게 수정해야할지 피드백을 정확히 이해하지 못해서 다시 한번 문의드립니다ㅠㅠ

godrm replied:
네. 자판기 내부에 음료수 별로 재고를 파악해보자는거죠
그래서 Dictionary<Beverage, Int> 형태로 구현을 많이 하더군요.

torch-ray replied:
[Beverage:Int]형태로 리턴하는 메서드는 따로 있긴한데... 위 코드대로가 아니라 전체 재고를 받아서 해당 딕셔너리에서 카운트를 각 음료수의 label로 넘겨주도록 구현해야 할까요?

godrm replied:
아.. 별도로 있는거군요. 그럼 왜 따로 각각 호출하는거죠?

torch-ray replied:
코드를 구현할때 모든 재고를 받아서 각 음료수별로 label에 재고수량을 넘겨주는것보다, 위 코드대로 특정 음료수를 넘겨주면 재고에서 filter를 해서 원하는 음료수의 수만 받아서 처리하는 방식이 filter함수만 활용하면 되기 때문에 편하다(?)고 생각했습니다. 이런 방식으로 구현하면 문제가 생길 수 있을까요?

torch-ray replied:
코드를 구현할때 모든 재고를 받아서 각 음료수별로 label에 재고수량을 넘겨주는것보다, 위 코드대로 특정 음료수를 넘겨주면 재고에서 filter를 해서 원하는 음료수의 수만 받아서 처리하는 방식이 filter함수만 활용하면 되기 때문에 편하다(?)고 생각했습니다. 이런 방식으로 구현하면 문제가 생길 수 있을까요?

torch-ray replied:
메뉴 리스트 만들어서 label별 각 재고 음료 재고 반영하였습니다!

godrm commented on VendingMachineApp/VendingMachineApp/VendingMachineViewController.swift:
```diff
@@ -33,66 +30,73 @@ class VendingMachineViewController: UIViewController {
     
     @IBAction func buttonForCantataStock(_ sender: Any) {
-        machine.addBeverage(beverage: factory.cantata)
+        vendingMachine.addBeverage(beverage: factory.cantata)
```
factory 코드가 diff에 안 보이는데, 이렇게 하면 추가할 때마다 새로운 음료 인스턴스가 생기는 건가요? 아니면 계속 같은 인스턴스를 추가하는 건가요?

torch-ray replied:
diff에 안보인다는 말을 이해하지 못했습니다 ㅠㅠ 설명부탁드립니다! 이 코드대로 추가하면 factory에 프로퍼티 중 하나(음료수 하나)를 가져와서 자판기에 추가하는 방식입니다!

godrm replied:
아.. 깃헙에서 변경된 파일만 보이고 변경이 없으면 diff에서 파일이 안보이거든요. 따로 찾아봐야해요. 
저는 바뀐 부분들만 위주로 보게 됩니다.

torch-ray replied:
아.. 팩토리 클래스는 이후 수정을 마치고 push할때 표시하겠습니다! ㅎㅎ

godrm reviewed (changes requested):
다음 피드백을 확인해서 개선해보세요.

torch-ray committed:
BeverageFactory Class 수정

NSCoding 과 Codable의 차이 및 다형성 학습 중 수정 필요한 부분 발견하여 Beverage로 통일해서 만들었던 음료를 각 Class타입으로 수정

torch-ray committed:
NSObject, NSCoding 프로토콜 추가 채택

NSCoding과 Codable의 차이점 학습 후 NSCoding채택 후 코드 변경, NSObject를 채택함에 따라 Hashable 프로토콜 삭제

torch-ray committed:
Codable 프로토콜 걷어내고 NSObject, NScoding으로 모두 수정

NSCoding은 Class만 사용가능하여 VendingMachine구조체 Class로 변경, 그외 Archiving과정에서 에러가 발생하여 enumType -> String으로 수정

torch-ray committed:
싱글톤 적용하여 싱글톤 객체로 접근하도록 수정

torch-ray commented:
> 다음 피드백을 확인해서 개선해보세요.

질문거리에 남긴것처럼  신딜리에 변수를 선언하고 뷰컨트롤러에서 접근하면 nil이 반환되는 결과에 대해 앱딜리와 신딜리의 차이점을 공부중인데, 단순히 앱딜리와 신딜리의 차이점에 대해 공부해서는 답을 찾기 어렵습니다... 혹시 관련하여 키워드나 힌트를 주실 수 있을까요?

godrm commented:
> > 다음 피드백을 확인해서 개선해보세요.
> 
> 질문거리에 남긴것처럼 신딜리에 변수를 선언하고 뷰컨트롤러에서 접근하면 nil이 반환되는 결과에 대해 앱딜리와 신딜리의 차이점을 공부중인데, 단순히 앱딜리와 신딜리의 차이점에 대해 공부해서는 답을 찾기 어렵습니다... 혹시 관련하여 키워드나 힌트를 주실 수 있을까요?

생성 시점과 참조 관계 때문입니다. 앱 델리게이트 - 신 델리게이트 - 뷰 컨트롤러가 각각 언제 생성되거 어떤 순서로 생성되는지 살펴보세요.

torch-ray commented:
감사합니다!

torch-ray committed:
편의 이니셜라이저로 각 음료에서 초기화값 정해주고 뷰컨트롤러에서 factory.으로 접근하는 코드 수정

torch-ray committed:
코드 변경에 따라 작동하지 않는 테스트 코드 삭제

godrm commented:
몇 가지 구조만 개선하고 머지하고 다음 단계를 하면서 개선해보는 게 좋겠습니다.
확인하고 댓글 남겨주세요.

torch-ray committed:
MenuList생성하여 Label에 각 음료별 재고 반영하도록 코드 수정

godrm commented:
수고하셨습니다.
